### PR TITLE
Add site-effect correction

### PR DESCRIPTION
**Pull request by claraElk**
Related to #132 

Few changes made so far:
- I added an optional flag to indicate that there are multiple sites in the phenotype file (`--site_correction`, default: False).
- If site correction is enabled, the site column is required to be named "site" and an error is raised otherwise (similar to the checks for the age/gender columns).
- So far, I included site as a covariate for QC-FC metric only (specified using this formula `"age + C(gender) + C(site)"`. We could also use dummy variables if that seems like a cleaner approach—what do you think?

We might need to update our test dataset to have the scanner column named "site" as required.

Todo:

- [x] Update test workflow to run with and without `--site_correction`
- [x] Add site correction in prediction metrics
- [x] Update test dataset phenotype file (with datalad)

### CODE REVIEW

**Comment by htwangtw** [comment]
Looks good so far  - let me know if you need help with the regression for site correction in the prediction part

**Comment by claraElk** [comment]
For the site correction for the prediction metrics, I had to change the way we were doing the cross-validation. Instead of using the function from scikit-learn `cross-validate` I added a loop to do within the cross-validation the site correction for the training and testing set separately (using a linear regression).

Let me know what you think of this approach.

File: wonkyconn/features/age_sex_prediction.py [hunk 1]
```diff
@@ -10,14 +10,53 @@
...
+
+    design_test = np.column_stack([np.ones(len(test_site)), test_site.to_numpy()])
+
+    beta, *_ = np.linalg.lstsq(
+        design_train,
+        X_train,
+        rcond=None,
+    )
+
+    beta_site = beta[1:, :]
```
Comment (htwangtw) [inline wonkyconn/features/age_sex_prediction.py]: You need to fit the linear regression on the training and testing sets separately
```python
beta_train, *_ = np.linalg.lstsq(design_train, X_train, rcond=None,)
beta_test, *_ = np.linalg.lstsq(design_test, X_test, rcond=None,)
X_train_corr = X_train - design_train[:, 1:] @ beta_train[1:, :]
X_test_corr = X_test - design_test[:, 1:] @ beta_test[1:, :]
```

File: wonkyconn/features/age_sex_prediction.py [hunk 2]
```diff
@@ -10,14 +10,53 @@
...
+    X_train_corr = X_train - design_train[:, 1:] @ beta_site
+    X_test_corr = X_test - design_test[:, 1:] @ beta_site
```
Comment (htwangtw) [inline wonkyconn/features/age_sex_prediction.py]: see above

File: wonkyconn/features/age_sex_prediction.py [hunk 3]
```diff
@@ -68,27 +110,53 @@ def training_pipeline(
...
+                y_prob = pipe.predict_proba(X_test)[:, 1]
+
+                scores["roc_auc"].append(roc_auc_score(y_test, y_prob))
+
+            else:
+                scores["mae"].append(mean_absolute_error(y_test, y_pred))
+
+                scores["r2"].append(r2_score(y_test, y_pred))
+
+    scores_df = pd.DataFrame(scores)
```
Comment (htwangtw) [inline wonkyconn/features/age_sex_prediction.py]: Check if the header of this and the old outputs are the same for compatibility if we are not cleaning this up by intergrating everything in sklearn pipeline function

**Review by htwangtw — CHANGES_REQUESTED** [review]
A few things - 
1. The function forgot to fit each site separately.
2. It would be clearer if we could create a customised function to fit into the sklearn pipeline function for cross-validation. There's [`FunctionTransformer`](
https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.FunctionTransformer.html) to create a sklearn-compatible object for the pipeline. This way you can also drop the duplication in `regress_site`

**Comment by claraElk** [comment]
Thanks for the feedback! I added the fitting step for each train/test set separately. 

I agree that creating a scikit-learn–compatible pipeline object would make the implementation much cleaner. I'll refactor the code as soon as the current implementation is approved.

File: wonkyconn/features/age_sex_prediction.py (hunk 3 above)
Comment (claraElk) [reply wonkyconn/features/age_sex_prediction.py]: Thanks! I did check that the headers were the same. 

But I don't think this will be an issue since I'm planning to wrap everything into a scikit-learn–compatible object (as you suggested above). Once that's done, the final implementation will likely return to the previous logic.

**Comment by htwangtw** [comment]
@claraElk I think we can merge this so we have a record of a version that we know works. If the refactor doesn't pan out, we have something to fall back on. Do you feel confident implementing with a sklearn function transformer?

**Comment by claraElk** [comment]
I am working on it today. I'll let you know how it goes. So far I tried with FunctionTransformer, but I don't think I might work here (as we might need to extract the beta only from our training set, and use these values for the test set; see my question below).  I am trying to create a class instead (based on @pbergeret12 code for another project). If I run into issues I'll let you know and we can merge this version instead.

Example of the class:
```
class SiteRegressor(BaseEstimator, TransformerMixin):
    """
     Regress out site effects from connectivity features.
    """

    def __init__(self, n_connectivity_features: int):
        self.n_connectivity_features = n_connectivity_features

    def fit(self, X, y=None):
        X_feat = X[:, :self.n_connectivity_features]
        site = X[:, self.n_connectivity_features:]
        
        # Add intercept
        design = np.column_stack([np.ones(len(site)), site])

        # Estimate coefficients on the training fold only
        self.beta_ = np.linalg.lstsq(design, X_feat, rcond=None)[0]

        return self

    def transform(self, X):
        X_feat = X[:, :self.n_connectivity_features]
        site = X[:, self.n_connectivity_features:]

        # Add intercept
        design = np.column_stack([np.ones(len(site)), site])

        # Remove only the site contribution (keep the intercept)
        return X_feat - design[:, 1:] @ self.beta_[1:]
```

However, I have a question regarding the current implementation. At the moment, we are performing two separate linear regressions (one on the training set and one on the testing set). I was discussing this with @pbergeret12, and it seems that we should estimate the beta values only on the training set and then apply these same coefficients to the testing set to avoid information leakage.

Would we need to modify the code before merging, for example:

```
beta_train, *_ = np.linalg.lstsq(design_train, X_train, rcond=None,)
X_train_corr = X_train - design_train[:, 1:] @ beta_train[1:, :]
X_test_corr = X_test - design_test[:, 1:] @ beta_train[1:, :] #?
```

What do you think @htwangtw ?

**Comment by htwangtw** [comment]
@pbergeret12, thank you for raising this point, and it got me to rethink my suggestion. Let's go with your suggestion now. Below is my train of thought:

You are correct about data leakage. If we treat the site as a property the model should learn and generalise to correct, then it makes sense to only fit on the training set. (Side note - for this to work, we need proper stratification.) 

However, we are working with an issue that doesn't have a real solution. The subject-per-site distribution can be way too imbalanced in neuroimaging, so we cannot really properly stratify, and leaving one site out just doesn't work. This leaves us with one option that is not correct and will potentially introduce a positive bias due to data leakage: regress out the site as a nuisance variable on the whole dataset before CV. This is the same logic as why I suggested doing the nuisance regression on both - I was treating it like a timeseries-level confound, but this doesn't really work. 

If we implement @pbergeret12's solution, we might get user complaints that things are not working, etc., depending on the nature of the dataset.

Implementing the second one will work more universally. It will be defensible but incorrect. And we will need to add a huge disclaimer to the documentation that no one would read.

Might be worth bringing this up to Lune and seeing what she thinks.

File: wonkyconn/features/age_sex_prediction.py [hunk 4]
```diff
@@ -10,14 +10,56 @@
...
+    )
+    beta_test, *_ = np.linalg.lstsq(
+        design_test,
+        X_test,
+        rcond=None,
+    )
+    X_train_corr = X_train - design_train[:, 1:] @ beta_train[1:, :]
+    X_test_corr = X_test - design_test[:, 1:] @ beta_test[1:, :]
+
+    return X_train_corr, X_test_corr
```
Comment (claraElk) [inline wonkyconn/features/age_sex_prediction.py]: Instead of this function I propose using this class to perform the fitting of the linear regression on the training set, and use these beta values for the test set as well: 

```python
class SiteRegressor(BaseEstimator, TransformerMixin):
    """
    Regress out site effects from connectivity features.
    """

    def __init__(self, n_connectivity_features: int):
        self.n_connectivity_features = n_connectivity_features

    def fit(self, X, y=None):
        X_feat = X[:, :self.n_connectivity_features]
        site = X[:, self.n_connectivity_features:]

        # Add intercept
        design = np.column_stack([np.ones(len(site)), site])

        # Estimate coefficients on the training fold only
        self.beta_ = np.linalg.lstsq(design, X_feat, rcond=None)[0]

        return self

    def transform(self, X):
        X_feat = X[:, :self.n_connectivity_features]
        site = X[:, self.n_connectivity_features:]

        # Add intercept
        design = np.column_stack([np.ones(len(site)), site])

        # Remove only the site contribution (keep the intercept)
        return X_feat - design[:, 1:] @ self.beta_[1:]
```

File: wonkyconn/features/age_sex_prediction.py [hunk 5]
```diff
@@ -68,27 +113,53 @@ def training_pipeline(
...
-        cv_results = cross_validate(
-            pipe,
-            connectivity_data,
-            y_train,
-            cv=cv_strategy,
-            scoring=scoring_metrics,
-            n_jobs=n_jobs,
-        )
-
-    scores_df = pd.DataFrame({k.replace("test_", ""): v for k, v in cv_results.items() if k.startswith("test_")})
```
Comment (claraElk) [inline wonkyconn/features/age_sex_prediction.py:81]: With the class function proposed above, we could go back to the previous workflow, and adding site correction directly within the scikit-learn pipeline: 

```python
    pipe = Pipeline(
        [
            ("imputer", SimpleImputer(strategy="median")),
            ("site_regression", SiteRegressor(connectivity_data.shape[1])),
            ("scaler", StandardScaler()),
            ("pca", PCA(n_components=n_pca, svd_solver="randomized", random_state=random_state,)),
            ("estimator", estimator),
        ]
    )

    with parallel_backend("threading", n_jobs=n_jobs):
        cv_results = cross_validate(
            pipe,
            X,
            y_train,
            cv=cv_strategy,
            scoring=scoring_metrics,
            n_jobs=n_jobs,
        )
```

**Comment by claraElk** [comment]
@htwangtw I implemented the code with @pbergeret12 suggestion for now. But I agress, we should discuss this in more details

**Review by htwangtw — APPROVED** [review]
We are missing tests for the new features. I would prefer to have it added at some point, either through a test dataset with the site as an argument or a unit test. @mathdugre @pbergeret12 are you guys okay with this direction? We can open an issue to log it.

File: wonkyconn/run.py [hunk 6]
```diff
@@ -66,6 +66,13 @@ def global_parser(exit_on_error: bool = True) -> argparse.ArgumentParser:
         default=False,
         help="Disable sex and age prediction to reduce runtime.",
     )
+    parser.add_argument(
+        "--site_correction",
```
Comment (HippocampusGirl) [inline wonkyconn/run.py]: I think this should be a dash instead of an underscore
```suggestion
        "--site-correction",
```

File: wonkyconn/workflow.py [hunk 7]
```diff
@@ -296,4 +304,8 @@ def load_data_frame(args: argparse.Namespace) -> pd.DataFrame:
         raise ValueError('Phenotypes file is missing the "gender" column')
     if "age" not in data_frame.columns:
         raise ValueError('Phenotypes file is missing the "age" column')
+    if getattr(args, "site_correction", True):
```
Comment (HippocampusGirl) [inline wonkyconn/workflow.py]: This should be equivalent in functionality but may be more readable. The site_correction attribute will always be set, because we define `default=False` above.
```suggestion
    if args.site_correction:
```

File: wonkyconn/config.py [hunk 8]
```diff
@@ -52,6 +53,7 @@ def from_cli_args(cls, args: argparse.Namespace | None) -> "WonkyConnConfig":
             debug=bool(getattr(args, "debug", False)),
             light_mode=bool(getattr(args, "light_mode", False)),
             suppress_warnings=bool(getattr(args, "suppress_warnings", False)),
+            site_correction=bool(getattr(args, "site_correction", False)),
```
Comment (HippocampusGirl) [inline wonkyconn/config.py]: The site_correction attribute will always be set, because we define default=False above. The `default` of `getattr` will never be used. It might also be a good idea to review this pattern in our CLI.
```suggestion
            site_correction=args.site_correction,
```

File: wonkyconn/features/age_sex_prediction.py [hunk 9]
```diff
@@ -18,6 +19,62 @@
...
+            self: Returns the instance itself.
+        """
+        connectivity_feat = connectivity_data_site[:, : self.n_connectivity_features]
+        site = connectivity_data_site[:, self.n_connectivity_features :]
+
+        # Add intercept
+        design = np.column_stack([np.ones(len(site)), site])
+
+        # Estimate coefficients on the training fold only
+        self.beta_ = np.linalg.lstsq(design, connectivity_feat, rcond=None)[0]
```
Comment (HippocampusGirl) [inline wonkyconn/features/age_sex_prediction.py]: Would it be more idiomatic and readable to use a `sklearn.linear_model.LinearRegression` here?

**Comment by HippocampusGirl** [comment]
Hi everyone,
I spent some adding a unit test for the prediction as well as the stratification by site for the cross validation. The code was mostly written by Fugu, Gemini and Claude, but I have spent some time refactoring it manually to improve readability. 
It works by generating data with and without site confounds, and then checking that the resulting prediction metrics are lower when we remove site confounds.
I also fixed some general type hints and completed the outstanding to dos that we talked about in the meeting. Making the smoke tests run both with and once without site correction doubles runtime, so we need to see if this is still tolerable.

**Comment by claraElk** [comment]
Hi @HippocampusGirl ! Looks good to me! 

I am currently testing the new implementation on additional datasets, and so far it seems to be working correctly.

Many thanks!

**Comment by claraElk** [comment]
Hi all!

We discussed with @HippocampusGirl the possibility of removing the gradients from the default metrics to save time and keep the workflow cleaner (i.e. we keep only the metrics we use, but we keep the possibility of running the gradients for future development). 

Also, I’ve tested the new site correction implementation on three datasets, and everything seems to be working as expected.

@htwangtw, do you think this is ready to be merged? Is there anything else you think we should add before merging?

Many thanks!

**Review by htwangtw — APPROVED** [review]
this is ready to merge!
